### ISSUE TITLE
Time limits and osqp_setup

### ISSUE THREAD

**Issue by joehuchette** [osqp/osqp#405]
Context: I encountered an instance with a O(1) second time limit and a very large scaling parameter value. Initializing via osqp_setup takes a very long time (far larger than the time limit), since AFAICT there are no checks against the time limit inside the function.

IMO it is a bit surprising that osqp_setup accepts a time limit parameter (namely, settings->time_limit) but does not check against it. This behavior also makes it difficult to cap the total "end-to-end" wall time inclusive of both set-up and solve.

What are your thoughts on adding logic that checks against the time limit inside osqp_setup? For my motivating instance, I would think to add a check to [this loop](https://github.com/osqp/osqp/blob/master/src/scaling.c#L72), though there are other spots in osqp_setup and functions it calls where such a check would be similarly appropriate (or not).

For reference, there's some existing discussion on this topic in #32.

**Comment by goulart-paul** [comment]
It seems to me that it should be very unusual for that scaling parameter to be large.    The loop you are citing there is performing a matrix equilibration procedure on the data that is a (I think slightly modified) implementation of the [Ruiz](https://epubs.siam.org/doi/pdf/10.1137/110825753) method.   That method generally converges very fast (see Thm 2.1 of the linked paper), and for most problems even just 3 iterations or so is sufficient.   From memory, we set the default to 10 just because the equilibration takes almost no time at all relative to everything else, and there were a couple of nasty examples in the Maros test set that required more.    Why do you want it be very large?    

If your problem data truly requires a huge number of equilibration iterations, then that seems indicative of some really bizarre scaling or ill-conditioning to start with.   If so, would it be possible to reformulate or rescale the problem data before passing to the solver instead?   
   
 Putting that aside, I'm not sure it makes sense to try to add further checks outside the main solver loop that halt execution when the time limit is reached.   The other places that come to mind during setup would be computing the AMD permutation, the elimination tree for the permuted KKT matrix and the initial KKT factorization.    The factorization is the most expensive by far so any further interrupts would be most likely to trigger there, but I don't know of any matrix factorization methods that support setting time limits in this way.    Note that the per-iteration cost of the Ruiz equilibration is about the same as the cost of multiplying the KKT matrix by a diagonal scaling, so far lower than the factorization cost.
 
 We could perhaps add a clarification to the documentation that the time limit really just means "don't attempt further iterations once this time has been reached".

**Comment by joehuchette** [comment]
I agree that instance I describe is a bit silly--I didn't write it by hand :)

But there are scenarios where you still might encounter instances like this. If you're running OSQP on a server, you might receive instances with memory corruption or that were crafted by an adversary, and you still want to be able to handle them gracefully.

Your point about the factorizations is well-taken. I guess I consider scaling a bit more dangerous/insidious since 1) it's not clear from the documentation that it doesn't consider the time_limit across scaling iterations, and 2) it is possible to increase this time spent on scaling (and, hence, outside the time_limit) by simply increasing a single parameter value, as opposed to something which scales with something more involved, like the problem size.

If you are interested, I can try to make a PR that checks against the time_limit before each scaling iteration. Regardless, I will probably just add a validation layer on my end that alters and/or rejects instances that seem troublesome. E.g. I could cap scaling at 1000 (or some other constant if you have a better suggested value, this is just a random guess). If you're interested, I could add this logic into OSQP proper as well.
